code/weapon/trails.cpp

```diff
@@ -190,129 +143,129 @@ void trail_render( trail * trailp )
 		sections[num_sections++] = n;
 	} while ( n != trailp->head );
 
-	if (num_sections <= 0)
+	if (num_sections <= 1)
 		return;
 
 	Assertion(ti->texture.bitmap_id != -1, "Weapon trail %s could not be loaded", ti->texture.filename); // We can leave this as an assert, but tell them how to fix it. --Chief
 
-	memset( &top, 0, sizeof(vertex) );
-	memset( &bot, 0, sizeof(vertex) );
-
-	// it's a tristrip, so allocate for 2+1
-	allocate_trail_verts((num_sections * 2) + 1);
-
 	float w_size = (ti->w_end - ti->w_start);
 	float a_size = (ti->a_end - ti->a_start);
 	int num_faded_sections = ti->n_fade_out_sections;
 
-	for (i = 0; i < num_sections; i++) {
+
+	vec3d prev_top, prev_bot; vm_vec_zero(&prev_top); vm_vec_zero(&prev_bot);
+	float prev_U = 0;
+	ubyte prev_alpha = 0;
+	for (int i = 0; i < num_sections; i++) {
 		n = sections[i];
+
+		// first get the alpha
 		float init_fade_out = 1.0f;
 
 		if ((num_faded_sections > 0) && (i < num_faded_sections)) {
 			init_fade_out = ((float) i) / (float) num_faded_sections;
 		}
 
-		w = trailp->val[n] * w_size + ti->w_start;
+		float w = trailp->val[n] * w_size + ti->w_start;
 
 		float fade = trailp->val[n];
 		
 		if (trailp->info.a_decay_exponent != 1.0f)
 			fade = powf(trailp->val[n], trailp->info.a_decay_exponent);
 
+		ubyte current_alpha = 0;
 		if (init_fade_out != 1.0f) {
```

Comment:
This if is not too nice, being a float equals comparison. Yes, we just set it to the same literal a few lines above so it probably will actually trigger despite being a float equals check, but that still doesn't make it pretty.
As far as I can tell, it should be easily doable to move the condition of the if that sets ``init_fade_out`` to be the condition of this if, and then only ever even declare ``init_fade_out`` _within_ this if's block.
Alternatively, this if could just be removed. It's not doing anything semantically, and is just there to keep two multiplications from happening. Though I am not even convinced that a branch is really cheaper than two float multiplications anyway.